### PR TITLE
[CARBONDATA-4104] Vector filling for complex decimal type needs to be handled

### PR DESCRIPTION
**Pull request by akkio-97**
### Why is this PR needed?
 Filling of vectors in case of complex decimal type whose precision is greater than 18 is not handled properly.
for ex- 
array<decimal(20,3)>
 
 ### What changes were proposed in this PR?
Ensured proper vector filling considering it's page data type.
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5290/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3530/

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java [hunk 1]
```diff
@@ -328,9 +328,29 @@ public BigDecimal getDecimal(Object valueToBeConverted) {
     public void fillVector(Object valuesToBeConverted, int size,
         ColumnVectorInfo vectorInfo, BitSet nullBitSet, DataType pageType) {
       CarbonColumnVector vector = getCarbonColumnVector(vectorInfo, nullBitSet);
-      //TODO handle complex child
-      int precision = vectorInfo.measure.getMeasure().getPrecision();
-      int newMeasureScale = vectorInfo.measure.getMeasure().getScale();
+      int precision;
```
Comment (VenuReddy2103) [inline core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:331]: These changes seem to be same as in`DecimalIntConverter.fillVector()`. Shall we reuse it instead of duplicating same logic again here ? Probably `DecimalUnscaledConverter` can extend `DecimalIntConverter` and overide fillVector like below ?

> public void fillVector(Object valuesToBeConverted, int size, ColumnVectorInfo vectorInfo, BitSet nullBitSet, DataType pageType) {
      if (valuesToBeConverted instanceof byte[][]) {
        ...
      } else {
        super.fillVector(valuesToBeConverted, size, vectorInfo, nullBitSet, pageType);
      }
    }

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java [hunk 2]
```diff
@@ -348,6 +368,32 @@ public void fillVector(Object valuesToBeConverted, int size,
             vector.putDecimal(i, value, precision);
           }
         }
+      } else if (valuesToBeConverted instanceof byte[]) {
```
Comment (VenuReddy2103) [inline core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:371]: Shall we avoid this code and reuse ? Same as above.

File: integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/SparkStoreCreatorForPresto.scala [hunk 3]
```diff
@@ -365,6 +365,15 @@ class SparkStoreCreatorForPresto extends QueryTest with BeforeAndAfterAll{
     sql(s"""LOAD DATA LOCAL INPATH '$resourcesPath/IUD/dest.csv' INTO TABLE streaming_table""")
   }
 
+  test("Test decimal unscaled converter") {
+    sql("drop table if exists array_decimal")
+    sql(
+      "CREATE TABLE IF NOT EXISTS array_decimal (salary array<decimal(20,3)>) STORED AS " +
+      "carbondata"
+    )
+    sql("insert into array_decimal select array(922.580, 3.435) ")
```
Comment (VenuReddy2103) [inline integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/SparkStoreCreatorForPresto.scala:374]: Didn't drop this table in `afterAll()` like other testcases do ?

File: integration/presto/src/test/scala/org/apache/carbondata/presto/integrationtest/PrestoTestUsingSparkStore.scala [hunk 4]
```diff
@@ -325,4 +326,22 @@ class PrestoTestUsingSparkStore
...
+  test("Test decimal unscaled converter") {
+    prestoServer
+      .execute(
+        "create table presto_spark_db.array_decimal(salary array(decimal(20,3)) ) with" +
+        "(format='CARBON') ")
+    copyStoreContents("array_decimal")
+    val result: List[Map[String, Any]] = prestoServer
+      .executeQuery("SELECT * FROM presto_spark_db.array_decimal")
+    assert(result.size == 1)
+    for (i <- 0 to 0) {
```
Comment (VenuReddy2103) [inline integration/presto/src/test/scala/org/apache/carbondata/presto/integrationtest/PrestoTestUsingSparkStore.scala:338]: result size is only 1. Suggest to remove this redundant loop if not required.

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java [hunk 5]
```diff
@@ -328,9 +328,29 @@ public BigDecimal getDecimal(Object valueToBeConverted) {
...
+      int newMeasureScale;
+      if (vectorInfo.measure == null) {
+        // complex primitive decimal flow comes as dimension
+        precision = ((DecimalType) vector.getType()).getPrecision();
+        newMeasureScale = ((DecimalType) vector.getType()).getScale();
+        size = ColumnVectorInfo.getUpdatedPageSizeForChildVector(vectorInfo, size);
+      } else {
+        precision = vectorInfo.measure.getMeasure().getPrecision();
+        newMeasureScale = vectorInfo.measure.getMeasure().getScale();
+      }
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:341]: can this be extracted into a common private method and reuse the same in all the places ?

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java (hunk 2 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:371]: yes, check and resue the code for this also

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3560/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5320/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5321/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3561/

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java (hunk 1 above)
Comment (akkio-97) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java]: done

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java (hunk 2 above)
Comment (akkio-97) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java]: done

File: integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/SparkStoreCreatorForPresto.scala (hunk 3 above)
Comment (akkio-97) [reply integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/SparkStoreCreatorForPresto.scala:375]: done

File: integration/presto/src/test/scala/org/apache/carbondata/presto/integrationtest/PrestoTestUsingSparkStore.scala (hunk 4 above)
Comment (akkio-97) [reply integration/presto/src/test/scala/org/apache/carbondata/presto/integrationtest/PrestoTestUsingSparkStore.scala]: done

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java [hunk 6]
```diff
@@ -88,11 +88,34 @@ void fillVector(Object valuesToBeConverted, int size, ColumnVectorInfo info, Bit
 
     DecimalConverterType getDecimalConverterType();
 
+    default int getPrecisionForDimension(CarbonColumnVector vector) {
```
Comment (Indhumathi27) [inline core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:91]: I can see, these methods are called from a single place only. Then, why it is required to create these methods ?. Instead, modify caller method like below:
"""
public void initializeValues(int size, CarbonColumnVector vector, ColumnVectorInfo vectorInfo) {
      if (vectorInfo.measure == null) {
        DecimalType decimalType = (DecimalType) vector.getType();
        // complex primitive decimal flow comes as dimension
        this.precision = decimalType.getPrecision();
        this.newMeasureScale = decimalType.getScale();
        this.updatedSize = ColumnVectorInfo.getUpdatedPageSizeForChildVector(vectorInfo, size);
      } else {
        CarbonMeasure carbonMeasure = vectorInfo.measure.getMeasure();
        this.precision = carbonMeasure.getPrecision();
        this.newMeasureScale = carbonMeasure.getScale();
      }
    } """

File: integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/SparkStoreCreatorForPresto.scala [hunk 7]
```diff
@@ -365,6 +366,15 @@ class SparkStoreCreatorForPresto extends QueryTest with BeforeAndAfterAll{
     sql(s"""LOAD DATA LOCAL INPATH '$resourcesPath/IUD/dest.csv' INTO TABLE streaming_table""")
   }
 
+  test("Test decimal unscaled converter") {
+    sql("drop table if exists array_decimal")
+    sql(
+      "CREATE TABLE IF NOT EXISTS array_decimal (salary array<decimal(20,3)>) STORED AS " +
```
Comment (Indhumathi27) [inline integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/SparkStoreCreatorForPresto.scala:372]: can you add a struct column with decimal type also

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java [hunk 8]
```diff
@@ -141,100 +165,116 @@ public void fillVector(Object valuesToBeConverted, int size,
...
-            if (value.scale() < newMeasureScale) {
-              value = value.setScale(newMeasureScale);
+      } else if (valuesToBeConverted instanceof byte[]) {
+        byte[] data = (byte[]) valuesToBeConverted;
+        if (pageType == DataTypes.BYTE) {
+          for (int i = 0; i < size; i++) {
+            if (nullBitSet.get(i)) {
+              vector.putNull(i);
+            } else {
+              BigDecimal value = BigDecimal.valueOf(data[i], scale);
```
Comment (Indhumathi27) [inline core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:193]: I think, we can still refactor the code for Cases: Byte, Short, Short_Int, Int and Long, As only difference is getting Bigdecimal value.

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/5051/

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java (hunk 8 above)
Comment (akkio-97) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:193]: Doing that might lead to NPE if Bigdecimal value is not initialized in any of the mentioned cases

File: integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/SparkStoreCreatorForPresto.scala (hunk 7 above)
Comment (akkio-97) [reply integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/SparkStoreCreatorForPresto.scala:372]: can add but it is of no use as the flow will be common for both

Comment (Indhumathi27) [reply integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/SparkStoreCreatorForPresto.scala:372]: better to add, as it is a new scenario without testcase

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java (hunk 8 above)
Comment (Indhumathi27) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:177]: i think, it will not lead to NPE, if all cases are handled. Please check

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3604/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5364/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5365/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3605/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5367/

**Comment by CarbonDataQA2** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3607/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5368/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3608/

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java (hunk 8 above)
Comment (akkio-97) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java]: done

File: integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/SparkStoreCreatorForPresto.scala (hunk 7 above)
Comment (akkio-97) [reply integration/spark/src/test/scala/org/apache/carbondata/integration/spark/testsuite/dataload/SparkStoreCreatorForPresto.scala:373]: okay

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java (hunk 6 above)
Comment (akkio-97) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java]: okay

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java (hunk 5 above)
Comment (akkio-97) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java]: done

**Comment by Indhumathi27** [comment]
LGTM

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java [hunk 9]
```diff
@@ -324,33 +311,6 @@ public BigDecimal getDecimal(Object valueToBeConverted) {
...
-          } else {
-            BigInteger bigInteger = new BigInteger(data[i]);
-            BigDecimal value = new BigDecimal(bigInteger, scale);
-            if (value.scale() < newMeasureScale) {
-              value = value.setScale(newMeasureScale);
-            }
-            vector.putDecimal(i, value, precision);
-          }
-        }
-      }
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:351]: add an else check and call the super.fillvector if it is not byte[][], instead of copying this code in super class and changing its implementation. super class was meant only for one dimensional byte[]

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java [hunk 10]
```diff
@@ -141,72 +151,48 @@ public void fillVector(Object valuesToBeConverted, int size,
...
-            }
-            vector.putDecimal(i, value, precision);
-          }
-        }
-      } else if (pageType == DataTypes.SHORT_INT) {
-        int shortIntSizeInBytes = DataTypes.SHORT_INT.getSizeInBytes();
+      if (this instanceof DecimalUnscaledConverter && scale < newMeasureScale) {
+        scale = newMeasureScale;
+      }
+      if (valuesToBeConverted instanceof byte[][]) {
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:157]: revert this change as discussed.

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java [hunk 11]
```diff
@@ -109,29 +113,35 @@ public BigDecimal getDecimal(Object valueToBeConverted) {
...
-      int newMeasureScale;
-      if (vectorInfo.measure == null) {
-        // complex primitive decimal flow comes as dimension
-        precision = ((DecimalType) vector.getType()).getPrecision();
-        newMeasureScale = ((DecimalType) vector.getType()).getScale();
-        size = ColumnVectorInfo.getUpdatedPageSizeForChildVector(vectorInfo, size);
-      } else {
-        precision = vectorInfo.measure.getMeasure().getPrecision();
-        newMeasureScale = vectorInfo.measure.getMeasure().getScale();
+      initializeValues(size, vector, vectorInfo);
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:137]: If used only once, no need to extract common code.

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5376/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3616/

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java (hunk 9 above)
Comment (akkio-97) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:351]: okay

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java [hunk 12]
```diff
@@ -327,25 +287,28 @@ public BigDecimal getDecimal(Object valueToBeConverted) {
...
-        for (int i = 0; i < size; i++) {
-          if (nullBitSet.get(i)) {
-            vector.putNull(i);
-          } else {
-            BigInteger bigInteger = new BigInteger(data[i]);
-            BigDecimal value = new BigDecimal(bigInteger, scale);
-            if (value.scale() < newMeasureScale) {
-              value = value.setScale(newMeasureScale);
+      if (valuesToBeConverted instanceof byte[]) {
+        super.fillVector(valuesToBeConverted, size, vectorInfo, nullBitSet, pageType);
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:291]: use return and remove else.

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java [hunk 13]
```diff
@@ -110,8 +110,8 @@ public BigDecimal getDecimal(Object valueToBeConverted) {
     }
 
     @Override
-    public void fillVector(Object valuesToBeConverted, int size,
-        ColumnVectorInfo vectorInfo, BitSet nullBitSet, DataType pageType) {
+    public void fillVector(Object valuesToBeConverted, int size, ColumnVectorInfo vectorInfo,
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:113]: revert this style change

File: core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java (hunk 12 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/metadata/datatype/DecimalConverterFactory.java:291]: could have merged into single statement. no need to modify now.  it's enough

**Comment by ajantha-bhat** [comment]
LGTM

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12444/job/ApacheCarbonPRBuilder2.3/5379/

**Comment by CarbonDataQA2** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12444/job/ApacheCarbon_PR_Builder_2.4.5/3619/
